File: infrastructure/src/test/java/com/study/platform/global/support/AbstractIntegrationTest.java
package com.study.platform.global.support;

import com.study.platform.domain.post.document.PostSearchRepository;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.MySQLContainer;

@SpringBootTest
@ActiveProfiles("test")
public abstract class AbstractIntegrationTest {

    @MockitoBean
    protected PostSearchRepository postSearchRepository;

    static final MySQLContainer<?> mysql;
    static final GenericContainer<?> redis;

    static {
        mysql = new MySQLContainer<>("mysql:8.0.36")
                .withDatabaseName("platform_test")
                .withUsername("test")
                .withPassword("test");
        redis = new GenericContainer<>("redis:7.4")
                .withExposedPorts(6379);

        mysql.start();
        redis.start();
    }

    @DynamicPropertySource
    static void overrideProperties(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", mysql::getJdbcUrl);
        registry.add("spring.datasource.username", mysql::getUsername);
        registry.add("spring.datasource.password", mysql::getPassword);
        registry.add("spring.data.redis.host", redis::getHost);
        registry.add("spring.data.redis.port", () -> redis.getMappedPort(6379));
    }
}

File: infrastructure/src/test/kotlin/com/study/platform/global/support/AbstractIntegrationTest.kt
package com.study.platform.global.support

import com.study.platform.domain.post.document.PostSearchRepository
import com.study.platform.support.AbstractContainerSupport
import jakarta.persistence.EntityManager
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.bean.override.mockito.MockitoBean

@SpringBootTest
@ActiveProfiles("test")
abstract class AbstractIntegrationTest : AbstractContainerSupport() {

    @Autowired
    protected lateinit var em: EntityManager

    @MockitoBean
    protected lateinit var postSearchRepository: PostSearchRepository
}
